### PR TITLE
Trade Probe

### PR DESCRIPTION
**Pull request by Kurfursten**
image

If
* There is no trader at all this shift
* The trade shuttle is in its starting location
* The trade port is open
* The accounts database is not in a state of financial mismanagement
* A trade probe has not visited

There is a chance for a trade probe to arrive in the trader dock with announcement. It contains a bunch of items from variety packs.
* Events range in weight. Some are rare (15 = meteors, rod), common (30 = ion storm, brand intelligence), very common (40 = carp, 50 = spam)
* Good finances give bonus weight. For every 3% everyone earned above their wages due to surplus budget, increase weight by 1, up to 30 weight at 190% wages.
* More crew gives bonus weight. For every 1 crewman above 10, increase weight by 3, up to 30 weight at 20 players.
* More time into the shift gives bonus weight. For every 2 minutes into the shift, increase weight by 1, up to 30 weight at 60 minutes.
* Base weight: 0 (max weight 90). Example: there are 24 players. Normal wage payout. 32 minutes into shift. Weight: 46

Items that appear in the pod come from random variety packs. All proceeds go to the trader account. After 9 minutes, a warning announcement plays. After 1 more minute, the probe docks at the vox outpost (salvage shuttle aux dock). If a trader latejoins, they can retrieve all the surplus items from the vendor.

🆑 
* rscadd: Automated trade probes may visit your system...

### CODE REVIEW

**Comment by ghost** [comment]
I like it. Should make the trade probe a little more soulful though, right now it just looks like a brown blob.

**Comment by Kurfursten** [comment]
@malfunctioning-ai We have a very kind offer from @BurntDevil to develop special Vox shuttle floors and walls in the next week for this PR

**Comment by GreatBigFailure** [comment]
I haven't checked the code, maybe you already have something like this in place, but some suggestions:
- Make the prices almost guaranteed to be higher than a trader would offer them at. Like 100% markup.
- Have it prioritize less commonly purchased items, not enough to cripple it, but enough to make it objectively worse than a player trader
- Have it prioritize upgrade injectors if there's no science, and ore boxes if there's no miner, etc, since this is a bigger benefit to dead/low pop
- Add an alert when someone signs up as a trader to tell them a probe went out. Maybe even include what was purchased.

I like the idea a lot, I just don't want it to interfere with player traders.

**Comment by Kurfursten** [comment]
> Make the prices almost guaranteed to be higher than a trader would offer them at. Like 100% markup.

This is generally the case. The probe only sells items from variety packs (e.g.: wonderful wardrobe, random chemicals, alcatraz iv) and sells items for 40-60% of the cost of the crate itself. So for example, the alcatraz crate costs $160. So to buy the baton helmet, it would cost you $64 to $96. All of these profits go to the trader account.

> Have it prioritize less purchased items... so it is worse than a player

Since it's random and it doesn't take requests or get better over time, it's definitely worse than a trader. It is going to naturally be a real mix of good and meh since it's pulling from variety packs.

> Prioritize certain products based on crew being present

I think this is too elaborate for the probe. This kind of logic is stuff real traders do.

> Add an alert

There are two things this does:
1. If someone joins while the trade probe is still at the station, the normal trade shuttle is locked and gives a message about the probe being in the dock. It also tells them exactly how long (in seconds) until the probe begins to depart. Most traders need at least 10 minutes to set up anyway.
2. If a player joins after the probe has left, it queues up a message at the trade window. The next time the player is greeted at the window, the window will inform them that a trade probe has already visited the station and they should check around the auxiliary dock to recover wares from it.

**Comment by Kurfursten** [comment]
Thank you to BurntDevil for some great icons

**Comment by Blithering** [comment]
Is there any security on the probe? Or can assistants smash the machine in hopes of free loot

**Comment by Kurfursten** [comment]
Like usual trade vendors, this one can't be hacked or otherwise looted

**Comment by Blithering** [comment]
Can it be unanchored? Also can trader IDs unlock it like the default tradevend or is it a custom one?

**Comment by GreatBigFailure** [comment]
> Like usual trade vendors, this one can't be hacked or otherwise looted

Didn't code dive, might already be the case but remember to make it not mechanic scannable.

**Comment by Kurfursten** [comment]
> Can it be unanchored? Also can trader IDs unlock it like the default tradevend or is it a custom one?

It's just like a normal tradevend, you can't unanchor it unless you have the trader debit card in your hand. Trader ID unlocks it as normal so you can recover the goods after you latejoin.

> remember to make it not mechanic scannable.

Actually a good catch

**Comment by GreatBigFailure** [comment]
> Actually a good catch

It is the first thing I look for, and only about half the time for nefarious purposes

File: code/modules/events/vox_trade_probe.dm [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+/datum/event/tradeprobe
+	endWhen	= 270 //How many 2-second ticks should we stay? 9 minutes = 540 / 2 = 270
+	oneShot = TRUE //one per shift
+
+/datum/event/tradeprobe/can_start(var/list/active_with_role)
+	//No probe if... there is a trader, the trade shuttle is at the station, or financial mismanagement
+	if(SStrade.loyal_customers.len)
+		return FALSE
+	if(!istype(trade_shuttle.current_port,/obj/docking_port/destination/trade/start))
```
Comment (ShiftyRail) [inline code/modules/events/vox_trade_probe.dm]: Should also check if a custom shuttle isn't at the trading port currently. Although the collision would be funny, I guess.

File: code/modules/events/vox_trade_probe.dm [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		command_alert(/datum/command_alert/tradeaversion_closedport)
+		return FALSE
+	if(payroll_modifier < 1)
+		command_alert(/datum/command_alert/tradeaversion_mismanagement)
+		return FALSE
+
+	//Gain 1 weight per 3% above normal wages players received last cycle, up to 30 weight at 190%
+	var/wageboost = min((payroll_modifier-1)/0.03,30)
+	//Gain 5 weight per player beyond 10, up to 30 weight at 20 players
+	var/playerboost = min(max(0,active_with_role["Any"]-10) * 3, 30)
```
Comment (ShiftyRail) [inline code/modules/events/vox_trade_probe.dm]: `Clamp(x, min, max) ?

File: code/modules/events/vox_trade_probe.dm [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+			TW.new_pending(tw_probe_return)
+
+/obj/machinery/vending/sale/trader/probe
+	anchored = TRUE
+	mech_flags = MECH_SCAN_FAIL
+
+/obj/machinery/vending/sale/trader/probe/New()
+	..()
+	//Grab some items from variety packs...
+	for(var/i = 1 to 6)
```
Comment (ShiftyRail) [inline code/modules/events/vox_trade_probe.dm]: `6` should be a define so that a new coder can easily change how many objects spawn, in my opinion.

**Review by ShiftyRail — COMMENTED** [review]
Not much to note other than - did you map in light on the trading probe ? Shuttles need them now.

**Comment by Kurfursten** [comment]
Yes, it has lights under each window and the door

**Comment by Kurfursten** [comment]
In my testing I have found that THIS event works perfectly but there are some problems with events themselves... I will need to fix them in a new PR

**Pull request merged**